pkg/proxy/benchmarking/benchmark.sh

```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/env bash
+
+# Copyright 2023 The Kubernetes Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -e
+set -o errexit
+set -o nounset
+set -o pipefail
+
+
+FILE=${0}
+RED="\033[91m"
+BLUE="\033[94m"
+ENDCOLOR="\033[0m"
+
+function help {
+    printf "\n"
+    printf "This script is expected to run twice for comparing initial and final benchmarks."
+    printf "\n\tFirst run for establishing a baseline benchmark this should be done by pointing"
+    printf "\n\tthe repository to master/main branch of the upstream."
+    printf "\n\tSecond run for generating benchmarks after code optimization and finally comparing "
+    printf "\n\tthe target with baseline benchmark."
+    printf "\n\nUsage: %s [-b] [-c count] [-f function] [-h]\n" "${FILE}"
+    printf "\t-b generate baseline benchmark, generates target benchmark if '-b' not specified.\n"
+    printf "\t-c count, run each benchmark n times.\n"
+    printf "\t-f function, regular expression for the function to run benchmark on.\n"
+    printf "\t-h help.\n"
+    printf "\nExample:\n\t %s -b -c 25 -f BenchmarkWriter\n" "${FILE}"
+    exit 1
+}
+
+function info_message(){
+    echo -e "[ ${BLUE}INFO${ENDCOLOR} ] ${1}"
+}
+
+function error_message(){
+    echo -e "[ ${RED}ERROR${ENDCOLOR} ] ${1}" >&2
+}
+
+function benchmark() {
+    mode=$1
+    count=$2
+    function=$3
+
+    info_message "mode: ${mode}"
+
+    info_message "running: go test --bench=${function} --benchmem --count=${count} --run=^# > ${mode}.report"
+
+    # generate comparison if the mode is target
+    if [[ "${mode}" == "target" ]]; then
+        # exit if baseline benchmark doesn't exist
+        if [[ ! -f "baseline.report" ]]; then
+            error_message "no baseline benchmark to compare with, run '${FILE} -b' to generate baseline benchmark for comparison\n"
+            exit 1
+        fi
+    fi
+
+    # run the benchmark
+    GOMAXPROCS=1 go test --bench="${function}" --benchmem --count="${count}" --run=^# > "${mode}.report"
```

Comment:
Why are we forcing GOMAXPROCS=1?

Reply:
I thought setting GOMAXPROCS=1 would ensure consistent env(single core only) for both the benchmarks - i.e. no multi-processing

Reply:
It won't enforce consistently having a full single core available or the same type of processor though and it seems like an unrealistic situation for kube-proxy?

Reply:
The only way we can ensure regularity is running the benchmark before / after on the same host.

Reply:
the plan is to run the benchmark before / after on same host only - https://github.com/kubernetes/kubernetes/issues/109481#issuecomment-1510197600

will remove GOMAXPROCS.